pkg/server/endpoints/authorized_entryfetcher.go

```diff
@@ -151,62 +228,113 @@ func (a *AuthorizedEntryFetcherWithEventsBasedCache) updateAttestedNodesCache(ct
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) updateAttestedNodeCache(ctx context.Context, spiffeID string) error {
+	node, err := a.ds.FetchAttestedNode(ctx, spiffeID)
+	if err != nil {
+		return err
 	}
 
+	// Node was deleted
+	if node == nil {
+		a.cache.RemoveAgent(spiffeID)
+		return nil
+	}
+
+	selectors, err := a.ds.GetNodeSelectors(ctx, spiffeID, datastore.RequireCurrent)
+	if err != nil {
+		return err
+	}
+	node.Selectors = selectors
+
+	agentExpiresAt := time.Unix(node.CertNotAfter, 0)
+	a.cache.UpdateAgent(node.SpiffeId, agentExpiresAt, api.ProtoFromSelectors(node.Selectors))
+
 	return nil
 }
 
-func buildCache(ctx context.Context, ds datastore.DataStore, clk clock.Clock) (*authorizedentries.Cache, uint, uint, error) {
+func buildCache(ctx context.Context, ds datastore.DataStore, clk clock.Clock) (*authorizedentries.Cache, uint, map[uint]time.Time, map[uint]time.Time, uint, error) {
 	cache := authorizedentries.NewCache(clk)
 
-	lastRegistrationEntryEventID, err := buildRegistrationEntriesCache(ctx, ds, cache, buildCachePageSize)
+	lastRegistrationEntryEventID, missedRegistrationEntryEvents, err := buildRegistrationEntriesCache(ctx, ds, clk, cache, buildCachePageSize)
 	if err != nil {
-		return nil, 0, 0, err
+		return nil, 0, nil, nil, 0, err
 	}
 
-	lastAttestedNodeEventID, err := buildAttestedNodesCache(ctx, ds, clk, cache)
+	lastAttestedNodeEventID, missedAttestedNodeEvents, err := buildAttestedNodesCache(ctx, ds, clk, cache)
 	if err != nil {
-		return nil, 0, 0, err
+		return nil, 0, nil, nil, 0, err
 	}
 
-	return cache, lastRegistrationEntryEventID, lastAttestedNodeEventID, nil
+	return cache, lastRegistrationEntryEventID, missedRegistrationEntryEvents, missedAttestedNodeEvents, lastAttestedNodeEventID, nil
 }
 
 // buildRegistrationEntriesCache Fetches all registration entries and adds them to the cache
-func buildRegistrationEntriesCache(ctx context.Context, ds datastore.DataStore, cache *authorizedentries.Cache, pageSize int32) (uint, error) {
-	lastEventID, err := ds.GetLatestRegistrationEntryEventID(ctx)
-	if err != nil && status.Code(err) != codes.NotFound {
-		return 0, fmt.Errorf("failed to get latest registration entry event id: %w", err)
+func buildRegistrationEntriesCache(ctx context.Context, ds datastore.DataStore, clk clock.Clock, cache *authorizedentries.Cache, pageSize int32) (uint, map[uint]time.Time, error) {
+	// Gather any events that may have been skipped during restart
+	var lastEventID uint
+	missedRegistrationEntryEvents := make(map[uint]time.Time)
+	resp, err := ds.ListRegistrationEntriesEvents(ctx, &datastore.ListRegistrationEntriesEventsRequest{})
+	if err != nil {
+		return 0, nil, err
+	}
+	for _, event := range resp.Events {
+		if event.EventID != lastEventID+1 && lastEventID != 0 {
+			for i := lastEventID + 1; i < event.EventID; i++ {
+				missedRegistrationEntryEvents[i] = clk.Now()
+			}
+		}
+		lastEventID = event.EventID
 	}
```

Comment:
We need to look back in the event history to see if there is any gaps because the event might commit after we get all entries.

Reply:
@MarcosDY  Yes, it is required.  When a long running write transaction that is modify SPIRE Entries doesn't commit because it is hung, it leaves gaps in the auto increment keys of the event table.  By tracking which event ids in the event table are missing, we can then go back and explicitly poll the skipped event ids prior to scanning the table for new entry event ids.  This is the best that can be done at the moment to ensure that we don't skip event ids when other transactions push the last seen event id beyond those that are stuck in hung database transactions.